cluster-autoscaler/cloudprovider/gce/gce_manager.go

```diff
@@ -381,12 +376,124 @@ func (m *gceManagerImpl) fetchAutoMigs() error {
 	}
 
 	if changed {
-		return m.cache.RegenerateInstancesCache()
+		return m.RegenerateInstancesCache()
 	}
 
 	return nil
 }
 
+// GetMigForInstance gets mig for given instance ref
+// Attempts to regenerate cache if there is a Mig with matching prefix in migs list.
+// TODO(aleksandra-malinowska): reconsider failing when there's a Mig with
+// matching prefix, but instance doesn't belong to it.
+func (m *gceManagerImpl) GetMigForInstance(instanceRef GceRef) (Mig, error) {
+	migRef, found := m.cache.GetMigRefForInstance(instanceRef)
+	if found {
+		mig, found := m.cache.getMig(migRef)
+		if !found {
+			return nil, fmt.Errorf("instance %+v belongs to unregistered mig %+v", instanceRef, migRef)
+		}
+		return mig, nil
+	}
+
+	for _, mig := range m.cache.GetMigs() {
+		migRef := mig.GceRef()
+
+		migBasename, found := m.cache.GetMigBasename(migRef)
+		var err error
+		if !found {
+			migBasename, err = m.GceService.FetchMigBasename(migRef)
+			if err != nil {
+				return nil, err
+			}
+			m.cache.SetMigBasename(migRef, migBasename)
+		}
+
+		if migRef.Project == instanceRef.Project &&
+			migRef.Zone == instanceRef.Zone &&
+			strings.HasPrefix(instanceRef.Name, migBasename) {
+
+			instanceRefs, err := m.regenerateInstanceCacheForMigReturnInstances(migRef)
+			if err != nil {
+				return nil, fmt.Errorf("error while looking for MIG for instance %+v, error: %v", instanceRef, err)
+			}
+
+			// see if instance found among instances obatained from GCE
+			found := false
+			for _, migInstanceRef := range instanceRefs {
+				if migInstanceRef == instanceRef {
```

Comment:
why not look it up in cache again?

Reply:
It could have been invalidated already.
I have some idea to add some "transactions" to cache interface. But it is more work and it is low prio. So I would skip it for now.